src/ngx_lua_resty_lmdb_module.c

```diff
@@ -289,26 +290,136 @@ static ngx_int_t ngx_lua_resty_lmdb_init_worker(ngx_cycle_t *cycle)
-                "unable to set LMDB encryption key, block size");
+                          "unable to set LMDB encryption key, block size");
             return NGX_ERROR;
         }
 
         rc = mdb_env_set_encrypt(lcf->env, lmdb_encrypt_func, &enckey, block_size);
         if (rc != 0) {
             ngx_log_error(NGX_LOG_CRIT, cycle->log, 0,
-                "unable to set LMDB encryption key: %s", mdb_strerror(rc));
+                          "unable to set LMDB encryption key: %s",
+                          mdb_strerror(rc));
             return NGX_ERROR;
         }
 
-        /* destroy data*/
-        ngx_explicit_memzero(lcf->key_data.data, lcf->key_data.len);
-        ngx_explicit_memzero(lcf->encryption_mode.data, lcf->encryption_mode.len);
+        /* worker will destroy secret data */
+        if (is_master == 0) {
+            ngx_explicit_memzero(lcf->key_data.data, lcf->key_data.len);
+            ngx_explicit_memzero(lcf->encryption_mode.data, lcf->encryption_mode.len);
+
+            ngx_str_null(&lcf->key_data);
+            ngx_str_null(&lcf->encryption_mode);
+        }
+    }
+
+    return NGX_OK;
+}
+
+
+static ngx_int_t
+ngx_lua_resty_lmdb_remove_files(ngx_cycle_t *cycle, ngx_path_t *path)
+{
+    ngx_file_info_t   fi;
+    ngx_err_t         err;
+
+    u_char            name_buf[NGX_MAX_PATH];
+    ngx_str_t        *names = LMDB_FILE_NAMES + 1;
+    ngx_str_t        *name;
+
+    if (ngx_file_info(path->name.data, &fi) == NGX_FILE_ERROR) {
+        ngx_log_error(NGX_LOG_CRIT, cycle->log, ngx_errno,
+                      ngx_file_info_n " \"%s\" failed", path->name.data);
+        return NGX_ERROR;
+    }
+
+    if (ngx_is_dir(&fi)) {
+
+        /* try to remove all lmdb files */
+        for (name = names; name->len; name++) {
+            ngx_snprintf(name_buf, NGX_MAX_PATH,
+                         "%V%V%Z", &path->name, name);
+
+            ngx_log_debug1(NGX_LOG_DEBUG_CORE, cycle->log, 0,
+                           "lmdb file remove: \"%s\"", name_buf);
+
+            if (ngx_delete_file(name_buf) == NGX_FILE_ERROR) {
+                ngx_log_error(NGX_LOG_WARN, cycle->log, ngx_errno,
+                              ngx_delete_file_n " \"%s\" failed",
+                              name_buf);
+            }
+        }
+
+        return NGX_OK;
+    }
+
+    ngx_lua_resty_lmdb_assert(!ngx_is_dir(&fi));
+
+    /* try to delete the file */
+    if (ngx_delete_file(path->name.data) == NGX_FILE_ERROR) {
+        ngx_log_error(NGX_LOG_WARN, cycle->log, ngx_errno,
+                      ngx_delete_file_n " \"%V\" failed",
+                      &path->name);
+    }
+
+    /* add a slash at the tail */
+    if (path->name.data[path->name.len - 1] == '/') {
+        ngx_snprintf(name_buf, NGX_MAX_PATH, "%V%Z", &path->name);
+
+    } else {
+        ngx_snprintf(name_buf, NGX_MAX_PATH, "%V/%Z", &path->name);
+    }
+
+    /* ensure lmdb directory exists */
+    err = ngx_create_full_path(name_buf, LMDB_DIR_MODE);
```

Comment:
This should not be needed. I think LMDB already creates the directory itself on `mdb_env_open`.

Reply:
Nginx will create the directory (ngx_conf_set_path_slot=> ngx_add_path=>cycle->paths) before lmdb module starts, so after we remove the file/directory we must re-create it again.